gateway/src/node/lp_listener/handshake.rs

```diff
@@ -0,0 +1,178 @@
+                    )));
+                }
+            }
+        }
+
+        // Continue handshake until complete
+        loop {
+            // Read incoming packet
+            let packet = self.receive_packet(stream).await?;
+
+            // Process the received packet
+            if let Some(action) = self
+                .state_machine
+                .process_input(LpInput::ReceivePacket(packet))
+            {
+                match action {
+                    Ok(LpAction::SendPacket(response_packet)) => {
+                        self.send_packet(stream, &response_packet).await?;
+                    }
+                    Ok(LpAction::HandshakeComplete) => {
+                        info!("LP handshake completed successfully");
+                        break;
+                    }
+                    Ok(other) => {
+                        debug!("Received action during handshake: {:?}", other);
+                    }
+                    Err(e) => {
+                        return Err(GatewayError::LpHandshakeError(format!(
+                            "Handshake error: {}",
+                            e
+                        )));
+                    }
+                }
+            }
+        }
+
+        // Extract the session from the state machine
+        self.state_machine.into_session().map_err(|e| {
+            GatewayError::LpHandshakeError(format!("Failed to get session after handshake: {}", e))
+        })
+    }
+
+    /// Send an LP packet over the stream with proper length-prefixed framing
+    async fn send_packet(
+        &self,
+        stream: &mut TcpStream,
+        packet: &LpPacket,
+    ) -> Result<(), GatewayError> {
+        use bytes::BytesMut;
+        use nym_lp::codec::serialize_lp_packet;
+
+        // Serialize the packet first
+        let mut packet_buf = BytesMut::new();
+        serialize_lp_packet(packet, &mut packet_buf).map_err(|e| {
+            GatewayError::LpProtocolError(format!("Failed to serialize packet: {}", e))
+        })?;
+
+        // Send 4-byte length prefix (u32 big-endian)
+        let len = packet_buf.len() as u32;
+        stream.write_all(&len.to_be_bytes()).await.map_err(|e| {
+            GatewayError::LpConnectionError(format!("Failed to send packet length: {}", e))
+        })?;
+
+        // Send the actual packet data
+        stream.write_all(&packet_buf).await.map_err(|e| {
+            GatewayError::LpConnectionError(format!("Failed to send packet data: {}", e))
+        })?;
+
+        stream.flush().await.map_err(|e| {
+            GatewayError::LpConnectionError(format!("Failed to flush stream: {}", e))
+        })?;
+
+        debug!(
+            "Sent LP packet ({} bytes + 4 byte header)",
+            packet_buf.len()
+        );
+        Ok(())
+    }
+
+    /// Receive an LP packet from the stream with proper length-prefixed framing
+    async fn receive_packet(&self, stream: &mut TcpStream) -> Result<LpPacket, GatewayError> {
```

Comment:
why are we duplicating packet serialisation/deserialisation (from `handler.rs`)? Can't we wrap `TcpStream` with a codec?

Reply:
Looking at it again, this is debt from moving from TcpStreams to one Tcp connection per packet (to remain sans-io-ish), it can be consolidated nicely